softioc/extension.c

```diff
@@ -85,33 +79,36 @@ void get_field_offsets(
     while (status == 0)
     {
         const char * field_name = dbGetFieldName(&dbentry);
-        int i;
-        for (i = 0; i < field_count; i ++)
-        {
-            if (strcmp(field_names[i], field_name) == 0)
-            {
-                field_offset[i] = dbentry.pflddes->offset;
-                field_size[i]   = dbentry.pflddes->size;
-                field_type[i]   = dbentry.pflddes->field_type;
-            }
-        }
+        PyObject *ost = Py_BuildValue("iii",
+            dbentry.pflddes->offset,
+            dbentry.pflddes->size,
+            dbentry.pflddes->field_type);
+        set_dict_item_steal(dict, field_name, ost);
         status = dbNextField(&dbentry, 0);
     }
 
     dbFinishEntry(&dbentry);
+    return dict;
 }
 
 
 /* Updates PV field with integrated db lookup.  Safer to do this in C as we need
  * an intermediate copy of the dbAddr structure, which changes size between
  * EPICS releases. */
-int db_put_field(const char *name, short dbrType, void *pbuffer, long length)
+static PyObject *db_put_field(PyObject *self, PyObject *args)
 {
+    const char *name;
+    short dbrType;
+    void *pbuffer;
+    long length;
+    if (!PyArg_ParseTuple(args, "shnl", &name, &dbrType, &pbuffer, &length))
+        return NULL;
+
     struct dbAddr dbAddr;
     int rc = dbNameToAddr(name, &dbAddr);
     if (rc == 0)
         rc = dbPutField(&dbAddr, dbrType, pbuffer, length);
-    return rc;
+    return Py_BuildValue("i", rc);
```

Comment:
Interesting issue here.  Turns out we ignore the return code at the point of call:
https://github.com/dls-controls/pythonIoc/blob/31fd6117d1d57f90273f75c77c16a2db67f63023/softioc/device.py#L205-L206
I guess we have two choices:
1. raise an exception instead of returning an ignored status code
2. Ignore the failure altogether

The calling function https://github.com/dls-controls/pythonIoc/blob/31fd6117d1d57f90273f75c77c16a2db67f63023/softioc/device.py#L194 is part of the user API, and so it would probably be an improvement to tell the caller that they've made a mistake, so I favour turning this `Py_BuildValue` return into setting an exception and returning `NULL` on failure.

Reply:
This now raises an exception
```
>>> t_ao.set(b"bad value")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/tom/Programming/pythonIoc/softioc/device.py", line 206, in set
    db_put_field(
  File "/home/tom/Programming/pythonIoc/softioc/imports.py", line 24, in db_put_field
    return _extension.db_put_field(name, dbr_type, pbuffer, length)
RuntimeError: dbPutField failed for TS-DI-TEST-01:AO
```